File: Chapter 1/src/Section1_2/gift1.java
/*
 ID: syk
 LANG: JAVA
 TASK: gift1
 */
package Section1_2;
	import java.util.*;
	import java.io.*;
public class gift1 {
	static Scanner in;
	static PrintWriter out;
	static int n;
	static ArrayList <String> name;
	static ArrayList <Integer> money;
	public static void main(String [] args) {
		try {
			in = new Scanner(new File("gift1.in"));
			out = new PrintWriter(new File("gift1.out"));
			
			init();
			out.println(solve());
			
			in.close();
			out.close();
		}
		catch(Exception e) {
			e.printStackTrace();
		}
	}
	static void init() {
		for(int i = 0; i < n; i++) {
			name.add(in.next());
			money.add(0);
		}
	}
	static String solve() {
		String ans = "";
		for(int i = 0; i < n; i++) {
			String givr = in.next();
			int mon = in.nextInt();
			int ppl = in.nextInt();
			int gm = mon/ppl;
			int rem = mon%ppl;
			int index = name.indexOf(givr);
			money.set(money.get(index), money.get(index)-mon+rem);
			for(int j = 0; j<ppl; j++) {
				String rec = in.next();
				index = name.indexOf(rec);
				money.set(money.get(index), money.get(index)+gm);
			}
			
			for(int j = 0; j<n; j++) {
				ans += name.get(j) + " " + money.get(j) + "\n";
			}
		}
		return ans;
	}
}

File: Chapter 1/src/Section1_2/gift1.py
fin = open("gift1.in", "r")
fout = open("gift1.out", "w")
n = int(fin.readline())

name = list()
mon = list()
for i in range(n):
    name.append(fin.readline())
    mon.append(0)
for i in range(n):
    giver = fin.readline()
    gi = name.index(giver)
    gift, ppl= map(int, fin .readline().split(" "))
    if ppl == 0:
        continue
    gm = gift//ppl
    mon[gi] = mon[gi] - gift + gift%ppl
    for j in range(ppl):
        rcvr = fin.readline()
        ri = name.index(rcvr)
        mon[ri] += gm
for i in range(n):
    fout.write(name[i][:-1] + " " + str(mon[i]) + "\n")
fin.close()
fout.close()